### ISSUE TITLE
mini.completion: Two-stage completion is not triggered by typing `.` with `rust-analyzer`

### ISSUE THREAD

**Issue by muni-corn** [echasnovski/mini.nvim#1230]
### Contributing guidelines

- [X] I have read [CONTRIBUTING.md](https://github.com/echasnovski/mini.nvim/blob/main/CONTRIBUTING.md)
- [X] I have read [CODE_OF_CONDUCT.md](https://github.com/echasnovski/mini.nvim/blob/main/CODE_OF_CONDUCT.md)
- [X] I have updated 'mini.nvim' to latest version

### Module(s)

mini.completion

### Description

At first I thought this was applicable to all languages and projects, but I was able to trigger two-step completion with `.` just fine in a TypeScript project. So this seems to apply to some languages and not others.

I spend most of my time in Rust projects where I have to wrestle with this issue. Oftentimes I'd like the completion window to pop up after typing `.` to see all of the methods available for a certain object. But when using `rust-analyzer`, this doesn't seem to work as expected.

### Neovim version

v0.10.1(Build type: Release)

### Steps to reproduce

Have `mini.completion` enabled and an LSP setup with `rust-analyzer`. (I can return with a minimal.lua if needed!)

For a minimal Rust project to test this, starting with a Rust toolchain:

1. `cargo new mini-test`
2. `cd mini-test`
3. `nvim src/main.rs`
4. You can replace the contents of main.rs with the following:

```rust
fn main() {
    let foo = String::from("Hello, world!");
}
```

Test autocompletion on `foo`.

1. On a new line below `let foo = ...`, type `foo.`.
2. Observe that two-step completion isn't triggered, or that we immediately fallback to simple completion.
3. Type any alphabetical character, like `p`. 
    a. If fallback completion was triggered before, use the hotkey to manually trigger two-step completion. It should suggest LSP completions as intended.
    b. If fallback completion was not triggered before, two-step completion should have been triggered automatically and should show LSP completions as intended.

You can also observe this behavior while trying to autocomplete modules.

1. On any new line, type `std::`
2. We should expect suggestions to show, but they don't. Repeat steps 2 and 3 from above.

### Expected behavior

Two-stage completion should be triggered after `.` is typed, so I can scroll through all available methods, fields, etc.

Example expected behavior in a JavaScript project: 
![20240918-131319-s](https://github.com/user-attachments/assets/f486304a-765f-44c1-83de-8b5f92746ac6)

### Actual behavior

Two-stage completion is not triggered, and the completion menu does not appear. I have to type another character for suggestions to appear.

In this Rust project, fallback completion seems to be triggered instead. (Sometimes it is, sometimes nothing is triggered at all.) Pressing <kbd>C-Space</kbd> to trigger two-step completion manually still falls back to basic completion.

![image](https://github.com/user-attachments/assets/2096d4c3-aca5-4a73-b214-0c6d0d3bd35e)

Apparently, this same behavior occurs when trying to autocomplete modules (trigger character `:`):

![image](https://github.com/user-attachments/assets/1fb0e67d-adb9-416b-9cb9-ee91cc3caa58)

but after typing a character, two-stage completion can be manually triggered as expected:

![image](https://github.com/user-attachments/assets/1781dfba-5cc2-42f9-89aa-168b9f1ca8ca)

![image](https://github.com/user-attachments/assets/4d697691-8ef7-4f94-9077-0818a59a8831)

**Comment by echasnovski** [comment]
Thanks for the very detailed issue! It was really helpful to provide the *exact* reproduction steps.

I did manage to reproduce most of these. For me pressing `<C-Space>` after typing `std::f` also didn't produce LSP suggestions (but fallback instead).

At first I thought that this is because of misconfigured triggers completion triggers characters (as in `server_capabilities.completionProvider.triggerCharacters`, but they seem to be fine. And it wouldn't explain LSP completion not appearing after `<C-Space>`.

Currently I suspect that this is an issue with LSP server, but I'll take a closer look.

**Comment by echasnovski** [comment]
I've finally found time to try and see what this can be caused by.

As far as I can tell, 'mini.completion' works as expected in every case here (after `foo.`, after `foo.p`, after `std::`):
- It sends proper 'textDocument/completion' request with proper context.
- It receives the result `{ isIncomplete = true, items = {} }`, meaning that LSP server says there are no completion candidates.

For what it is worth, I can observe the same behavior with 'hrsh7th/nvim-cmp' and 'hrsh7th/cmp-nvim-lsp' source. So I don't think it is a problem on 'mini.completion' side. I tried searching issues for this in 'rust-lang/rust-analyzer', but was not successful.

Closing as this looks like an upstream problem. If you'll have more details, feel free to post a comment here and we'll try to take a further look.
